samza-core/src/main/java/org/apache/samza/table/remote/RemoteTableDescriptor.java

```diff
@@ -0,0 +1,188 @@
+    }
+
+    // Serialize and store reader/writer rate limiters with config
+    if (readRateLimiter == null && maxReadRate > 0) {
+      readRateLimiter = new EmbeddedRateLimiter(maxReadRate);
+    }
+
+    if (readRateLimiter != null) {
+      tableSpecConfig.put(RemoteReadableTable.READ_RLIMIT, serializeObject("read rate limiter", readRateLimiter));
+    }
+
+    if (writeRateLimiter == null && maxWriteRate > 0) {
+      writeRateLimiter = new EmbeddedRateLimiter(maxWriteRate);
+    }
+
+    if (writeRateLimiter != null) {
+      tableSpecConfig.put(RemoteReadableTable.WRITE_RLIMIT, serializeObject("write rate limiter", writeRateLimiter));
+    }
+
+    return new TableSpec(tableId, serde, RemoteTableProviderFactory.class.getName(), tableSpecConfig);
+  }
+
+  /**
+   * Use specified TableReadFunction with remote table.
+   * @param readFn read function instance
+   * @return this table descriptor instance
+   */
+  public RemoteTableDescriptor<K, V> withReadFunction(TableReadFunction<K, V> readFn) {
+    Preconditions.checkNotNull(readFn, "null read function");
+    this.readFn = readFn;
+    return this;
+  }
+
+  /**
+   * Use specified TableWriteFunction with remote table.
+   * @param writeFn write function instance
+   * @return this table descriptor instance
+   */
+  public RemoteTableDescriptor<K, V> withWriteFunction(TableWriteFunction<K, V> writeFn) {
+    Preconditions.checkNotNull(writeFn, "null write function");
+    this.writeFn = writeFn;
+    return this;
+  }
+
+  /**
+   * Use specified rate limiter for table read operations
+   * @param limiter rate limiter for reads
+   * @return this table descriptor instance
+   */
+  public RemoteTableDescriptor<K, V> withReadRateLimiter(RateLimiter limiter) {
+    Preconditions.checkNotNull(limiter, "null read rate limiter");
+    this.readRateLimiter = limiter;
+    return this;
+  }
+
+  /**
+   * Use specified rate limiter for table write operations
+   * @param limiter rate limiter for writes
+   * @return this table descriptor instance
+   */
+  public RemoteTableDescriptor<K, V> withWriteRateLimiter(RateLimiter limiter) {
+    Preconditions.checkNotNull(limiter, "null write rate limiter");
+    this.writeRateLimiter = limiter;
+    return this;
+  }
+
+  /**
+   * Specify rate limit for table read operations (withReadRateLimiter takes precedence)
+   * @param maxRate rate limit for read operations; must be positive
+   * @return this table descriptor instance
+   */
+  public RemoteTableDescriptor<K, V> withMaxReadRate(int maxRate) {
+    Preconditions.checkArgument(maxRate > 0, "Max read rate must be a positive number.");
+    this.maxReadRate = maxRate;
+    return this;
+  }
+
+  /**
+   * Specify rate limit for table write operations (withWriteRateLimiter takes precedence)
+   * @param maxRate rate limit for write operations; must be positive
+   * @return this table descriptor instance
+   */
+  public RemoteTableDescriptor<K, V> withMaxWriteRate(int maxRate) {
```

Comment:
Makes sense, although I don't see any value of having a fixed number of credits per-request other than 1. I am guessing you're alluding to bytes/sec kind of throttling. The tough thing here is that we don't have the concept of the size of each table record. So this is not readily supportable. 

The getRateLimitTag in ReadFn and WriteFn were supposed to allow applications to provide the tags for each stream message. However, since I'm dropping them per recommendation from the ReadFn/WriteFn interfaces. We would have to come up with another pluggable piece to allow application to customze the tags. 

As pointed out in the code comment, the initial implementation is blocking only. This is also inline with the semantics of the IO functions right now. I think @weisong44 and I were thinking to address these in follow up PRs instead of tackle them in one big shot. 

Yeah, I've called a meeting on Monday. Maybe we can go over these points then.